File: julia/2018-10-10-Intro-Bayesian-Linear-Regression.jl
using DataFrames
using Distributions
using Plots
using Random
Random.seed!(111)

function simulate(n = 1000, w0 = -.3, w1 = -.7, alpha = 1 / 5)
    x = rand(Uniform(-1, 1), n)

    w = [w0; w1]
    A = [ones(size(x)) x]
    
    y = A * w + rand(Normal(0, alpha), size(x))
    return (y, x)
end

x, y = simulate()
println(x[1:10])
println(y[1:10])

p = plot(x, y, markershape = :circle, linealpha = 0)
display(p)
# n = 1000; α = 1 / 5;
# w₀ = -.3; w₁ = -.7;
# w = [w₀; w₁];

# x = rand(Uniform(-1, 1), n);
# A = [ones(size(x)) x];
# y = A * w + rand(Normal(0, α), size(x));

# p = plot(;x = x, y = y)
# draw(SVG(5inch, 4inch), p)

# β = (1 / 2)^2
# μ = zeros(2)
# Σ = Matrix{Float64}(I, 2, 2) * β

# w = rand(MvNormal(μ, Σ), 1)

# Σ⁻¹ = α * A'A + β * I;
# Σ = (Σ⁻¹)^(-1);
# μ = α * Σ * A' * y;

# df = DataFrame(x = x, y = y);
# for i in 1:1000
#     w = rand(MvNormal(μ, Σ))
#     df[Symbol("y" * string(i))] = A * w
# end

File: python/2018-10-10-Intro-Bayesian-Linear-Regression.py
import numpy as np

def simulate(n : int = 1000, w0 : float = -.3, w1 : float = -.7, alpha : float = 1 / 5) -> tuple:
    x = np.random.uniform(-1, 1, n)
    
    w = np.array([w0, w1])
    A = np.column_stack((np.ones(x.shape), x))
    
    y = A.dot(w) + np.random.normal(0, alpha, size = n)
    return (x, y, A, alpha)

def posterior(y : np.array, A : np.array, alpha : float, beta : float = (1 / 2)**2) -> tuple:
    sigmainv = alpha * A.T.dot(A) + beta * np.eye(2)
    sigma = np.linalg.inv(sigmainv)
    mu = alpha * sigma.dot(A.T).dot(y)
    return (mu, sigma)

x, y, A, a = simulate();
mu, sigma = posterior(y, A, a);